src/type_checkers/array_access.cr

```diff
@@ -0,0 +1,24 @@
+module Mint
+  class TypeChecker
+    # type_error NegatedExpressionNotBool
+
+    def check(node : Ast::ArrayAccess) : Type
+      lhs = node.lhs
+      index = node.index
+
```

Comment:
The left hand side needs to be checked in all cases (like this `type = check node.lhs`) because there can be type errors in them, this will give you the type of the expression or raise a type error which is handled outside of the type checker so you don't need to deal with it.

Then you can compare  and fill the type with `Array(a)` like this:

`Comparer.compare(type, Type.new("Array", [Type.new("a")]))` 

which if returns `nil` it means that the two types are not equal, in this case you need to raise a type error, otherwise it will return the actual type (where holes are filled), of which you can use the first parameter (because its an `Array`) as the first parameter of the `Maybe` which is returned.